megatron/core/pipeline_parallel/schedules.py

```diff
@@ -654,92 +673,116 @@ def backward_step_helper(microbatch_id):
 
     # Run warmup forward passes.
     parallel_state.set_virtual_pipeline_model_parallel_rank(0)
-    input_tensors[0].append(
-        p2p_communication.recv_forward(tensor_shape,
-                                       dtype=dtype,
-                                       batch_p2p_comm=batch_p2p_comm,
-                                       timers=timers))
+
+    # iterate through each num_pipeline_model_parallel_groups_to_track
+    for _ in range(num_pipeline_model_parallel_groups_to_track):
```

Comment:
Does the `_` here have same value as `_COMPONENT_CONNECTOR_GROUP_INDEX`?
And I guess you uses the global variable `_COMPONENT_CONNECTOR_GROUP_INDEX` here so the `forward_step_helper` and `backward_step_helper` can pick it up and use it without having an additional argument?

Reply:
Yes, you are correct, `_` has the same value as `_COMPONENT_CONNECTOR_GROUP_INDEX`. This is probably a consequence of how I designed this, I first defined `_COMPONENT_CONNECTOR_GROUP_INDEX` before defining all of the loops. We could also implement `_COMPONENT_CONNECTOR_GROUP_INDEX` as a variable passed through the helper functions rather than a global variable, but functionally I don't think there is a difference.

Reply:
sure I will add this as a precaution